Remove OIDC, add support for Django 2.2

bradenmacdonald opened the pull request:
Django 2.2 LTS was recently released, and this PR adds support for that while also removing support for Django versions 1.8 through 1.10 which [are no longer officially supported](https://www.djangoproject.com/download/#supported-versions).

This also bumps the versions of python-social-auth to get their django compatibility fixes.

I expect that most apps using this library will already be upgraded to Django. 1.11+

**New**: This PR also removes the long-deprecated OIDC provider. People who still need it can pin to `auth-backends<3.0`

Together these changes comprise a new version 3.0.0

bradenmacdonald commented:
@robrap @ormsbee I was looking into upgrading our Blockstore and LabXchange projects to Django 2.2 LTS and this repo seems to be the only blocker. Would either of you be able to help me with the TODO item mentioned above at some point? (Updating one test file to work with upstream JWT library changes)

Not at all urgent :)

mikix commented on auth_backends/__init__.py:
```diff
@@ -3,4 +3,4 @@
  projects as well.
 """
-__version__ = '2.0.0'  # pragma: no cover
+__version__ = '2.1.0'  # pragma: no cover
```
I would say bump to 3.0.0, since we are dropping supported platforms (ones that likely don't matter to anyone at this point, but still).

bradenmacdonald replied:
@mikix I've rebased this and changed the version to `3.0.0` as you suggested. Do you know if there's anyone who could help with fixing `test_backends.py` to work with the changes to the upstream OpenIdConnectTestMixin ?

mikix replied:
I tried to look at fixing them, and it was more complicated than I had time for. I'm trying to find someone at edX that would know more.

bradenmacdonald replied:
> I tried to look at fixing them, and it was more complicated than I had time for. 

Same here, haha. Thanks!

mikix commented on test_settings.py:
```diff
@@ -26,13 +26,12 @@
     'django.middleware.common.CommonMiddleware',
     'django.middleware.csrf.CsrfViewMiddleware',
     'django.contrib.auth.middleware.AuthenticationMiddleware',
-    'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
```
Why?

bradenmacdonald replied:
"The SessionAuthenticationMiddleware class is removed. It provided no functionality since session authentication is unconditionally enabled in Django 1.10." -> Django 2.0 release notes

nasthagiri commented:
@bradenmacdonald Thanks for this PR.  The tests that are failing are related to `EdXOpenIdConnect`.  `EdXOpenIdConnect` has been deprecated for some time in favor of EdXOAuth2.

The only remaining edX.org service that actively uses `EdXOpenIdConnect` is `edx-analytics-dashboard`.  At edX, we are very close to updating `edx-analytics-dashboard` to move over to `EdXOAuth2`.  To warn the community of this, we can create a DEPR ticket so they are aware of this removal as part of Koa - for Django-2.

So my recommendation is to just remove `EdXOpenIdConnect` before upgrading to Django-2.  I spoke to @feanil earlier today and we see value in finishing up our work to remove usages of the old DOP provider and `EdXOpenIdConnect` rather than upgrading them to Django-2.

bradenmacdonald commented:
> So my recommendation is to just remove EdXOpenIdConnect before upgrading to Django-2.

@nasthagiri That's a great idea! I'll update this PR so that version 3.0 introduces Django 2.x support, and removes `EdXOpenIdConnect` and django 1.8 through 1.10. (But it may take some time until I get that done, so someone else is welcome to beat me to it.)

nasthagiri commented:
Sounds good.  Thank you @bradenmacdonald.

bradenmacdonald commented:
@nasthagiri @mikix I have updated this to remove the OIDC provider, and have also updated some out-of-date info in the Readme. Would either of you be able to take a look and merge/release ?

nasthagiri commented:
@pwnage101 @doctoryes Do you see any issues with removing support for `EdXOpenIdConnect` from `auth-backends` in order to unblock Django-2 upgrades?

Doing an [edX-wide search](https://github.com/search?p=2&q=org%3Aedx+EdXOpenIdConnect&type=Code), I see we'll need to remove `EdXOpenIdConnect` as "backup" backends from `ecommerce` and `course-discovery`.  `analytics-dashboard` can continue to use a different branch of this repo while it's still migrating.

bradenmacdonald commented:
@nasthagiri @mikix Thanks - is anything else needed before we merge this?  It seems that `edx-analytics-dashboard` already pins the version of `edx-auth-backends` to an old release, so only `ecommerce` and `course-discovery` need to be changed. Does anyone at edX have a few minutes to make those updates? I can submit a quick PR to remove the one line from `AUTHENTICATION_BACKENDS` for each, but it seems like there are a bunch of other OIDC settings and test cases that may need some minor updates as well.

nasthagiri commented:
@mikix Yes, I did miss the emojis.  Thanks for the heads up.

> On Discovery side, removing it as a backup auth backend - is that like, a no-impact config change because it shouldn't normally be used, or is there also code changes that would have to go with it

Yes, it's only a config change at this point.  There should be no code changes.  The same is needed for ecommerce - also only config changes. 

@mikix As a rep from the Engagement theme (that owns both services), it would be great if you can get those changes in (through your teams).  If not, the Arch team will prioritize this work after the Py-3 upgrade.

Thanks all!

mikix commented:
Finally got around to making the disco-side PRs:
https://github.com/edx/course-discovery/pull/2306
https://github.com/edx/devstack/pull/455

Can look at changing ecommerce once disco goes well.

bradenmacdonald commented:
@mikix Thanks for those changes! Can you please take a look at ecommerce too when you get time, and then can you or @nasthagiri let me know if this PR is good to merge?

mikix commented:
Yeah, I'll do ecommerce versions too. The discovery side of things is just now landing all its little bits.

mikix commented:
Ecommerce PRs here: https://github.com/edx/ecommerce/pull/2731 (that has links to rest)

jlajoie commented:
We've merged in https://github.com/edx/ecommerce/pull/2731 to remove OIDC from ecommerce. This PR should be unblocked now

mikix commented on .travis.yml:
```diff
@@ -1,20 +1,21 @@
 
 python:
-  - 2.7
+  - 3.6
```
Small note, but I think most of OpenEdX is still on 3.5. And this packages's readme only mentions 3.5.

timmc-edx replied:
It looks like the rest of the config was already claiming 3.6, despite the README. I think that might call for a second PR—and I'm planning on putting one in to make this OEP-18 compliant, which might be a good time to make that change as well.

nasthagiri commented:
Thanks all!
@jmbowman Looks like this PR is now unblocked by upgrading the other IDAs.  Can you add review of this PR to the Arch team’s backlog?

jmbowman commented:
Added a ticket to the top of the backlog: https://openedx.atlassian.net/browse/BOM-1164

timmc-edx commented on auth_backends/backends.py:
```diff
@@ -23,211 +13,10 @@
-
 def _to_language(locale):
     """Convert locale name to language code if necessary.
 
```
The comment here could probably be changed to no longer mention OIDC. (Is it true for the OAuth side of things as well?)

feanil commented on setup.py:
```diff
@@ -36,10 +36,10 @@
         'six',
-        'social-auth-core[openidconnect]>=1.3.0,<2.0.0',
-        'social-auth-app-django>=1.2.0,<2.0.0',
+        'social-auth-core[openidconnect]>=3.1.0,<4.0.0',
```
+1 to removing this.

feanil commented on test_requirements.txt:
```diff
@@ -9,3 +9,6 @@ pytest-cov>=2.4.0,<3.0.0
 pytest-django>=3.1.2,<4.0.0
 tox>=2.3.2,<3.0.0
 unittest2>=1.1.0,<2.0.0
+python-jose>=3.0.0
```
Agree, It looks like this is only needed for OIDC.  I think it can be removed along with the above extra.

feanil commented on .travis.yml:
```diff
@@ -1,20 +1,21 @@
+  - python: 2.7
     env: TOXENV=django111
+  - env: TOXENV=quality
+  - env: TOXENV=django111
```
This line causes the django111 tests to run with all python versions listed above.  So we need it to get the py35-django111 tests to run.  Ideally we eventually remove the `py27-django111` include at some point and than this line can scale out to more python versions simply by adding them above.

timmc-edx replied:
After offline discussion: It looks like this is actually covered by the baseline env; I'll remove it, but make sure the expected Travis builds do in fact run.

bradenmacdonald commented:
Certainly feel free to pick this up. I will otherwise come around and finish it at some point, but don't know yet when I'll get to it.

timmc-edx commented:
Merged as https://github.com/edx/auth-backends/pull/68

edx-webhook commented:
Although this pull request is already closed, I've created [OSPR-5446](https://openedx.atlassian.net/browse/OSPR-5446) so that we can track it in Jira.

There is nothing you have to do. No action is needed from your side.  Thanks.

<!-- comment:welcome_closed -->
<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->
